### PR TITLE
New column and field types for managing permissions and roles with permissions

### PR DESCRIPTION
**Pull request by shaoshiva**
Hi,

This PR implements 2 new types of field and a column to display and manage permissions and roles with permissions in a more efficient way. These are intended to be used by the permission manager in replacement of the `checklist` and `checklist_dependency` types. I used the views of these ones as a code base, so there are similarities in the code, although I have refactored a lot.

I would have liked to implement these new fields and column directly in the permissions manager but I didn't found a clean way to implement a new field type from another package. I think this could be implemented with little changes in the code, but I will see another time, there are many other more interesting things to do in my opinion.

#### See in images

alt text &nbsp; alt text &nbsp; alt text

#### Main features
- groups of permissions by prefix (see #928) for fields and column
- buttons to check/uncheck all the permissions or a group of permissions
- permissions granted by a role no longer overwrite the permissions granted directly to the user, this is intended to avoid loosing permissions when checking and then unchecking a role. See below how it works now :
laravel-backpack-pr-field-type-role-permission-demo1

#### Usage

```php
$this->crud->setColumns([
    'label'     => trans('backpack::permissionmanager.extra_permissions'), // Table column heading
    'type'      => 'permissions',
    'name'      => 'permissions', // the method that defines the relationship in your Model
    'entity'    => 'permissions', // the method that defines the relationship in your Model
    //'attribute' => 'name', // foreign key attribute that is shown as the permission name (string, callback or null for default value)
    'model'     => config('laravel-permission.models.permission'), // foreign key model
    'inline'    => false, // true for all groups of permission on the same row or false for each on a row
]);

$this->crud->addField([
    // two interconnected entities
    'label'             => false,
    'field_unique_name' => 'user_role_permission',
    'type'              => 'permissions_with_roles',
    'name'              => 'roles_and_permissions', // the methods that defines the relationship in your Model
    'subfields'         => [
        'primary' => [
            'label'            => trans('backpack::permissionmanager.roles'),
            'name'             => 'roles', // the method that defines the relationship in your Model
            'entity'           => 'roles', // the method that defines the relationship in your Model
            'entity_secondary' => 'permissions', // the method that defines the relationship in your Model
            'attribute'        => null, // foreign key attribute that is shown to user (string, callback or null for default value)
            'model'            => config('laravel-permission.models.role'), // foreign key model
            'pivot'            => true, // on create&update, do you need to add/delete pivot table entries?]
            'columns'          => true, // Number of columns (1,2,3,4,6) or "true" for all on same line or "false" for each on a single line
        ],
        'secondary' => [
            'label'          => ucfirst(trans('backpack::permissionmanager.permission_singular')),
            'name'           => 'permissions', // the method that defines the relationship in your Model
            'entity'         => 'permissions', // the method that defines the relationship in your Model
            'entity_primary' => 'roles', // the method that defines the relationship in your Model
            'attribute'      => null, // foreign key attribute that is shown to user (string, callback or null for default value)
            'model'          => \Backpack\PermissionManager\app\Models\Permission::class, // foreign key model
            'pivot'          => true, // on create&update, do you need to add/delete pivot table entries?]
        ],
    ],
]);
```

Todos:
- [x] add a new column type to display permissions grouped by prefix 
- [x] sort groups of permissions in alphabetical order
- [ ] create a pull request to use these field types by default in the permission manager

### CODE REVIEW

**Comment by lloy0076** [comment]
Screenshots and idea seems sound - I'll keep an eye on this.

**Comment by shaoshiva** [comment]
Added sorting groups of permissions in alphabetical order and column to display permissions grouped by prefix.

**Comment by niladam** [comment]
Not that's really impressive and i think it could really improve backpack :)

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 1]
```diff
@@ -52,4 +46,32 @@ public function search()
...
     }
+
+    /**
+     * Gets the searchable columns.
+     *
+     * @return Collection
+     */
+    protected function searchableColumns()
+    {
+        // create an array with the names of the searchable columns
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: Not necessary (nor harmful).

File: src/resources/views/columns/permissions.blade.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
+{{-- relationships with pivot table (n-n) --}}
+<td>
+    <?php
+
+    // Gets the attribute to display the permission name
+    $attribute = array_get($column, 'attribute');
```
Comment (lloy0076) [inline src/resources/views/columns/permissions.blade.php]: If that were named something less generic than `$attribute` the comment above would be somewhat more obvious; or if `$attribute` really does describe what this is, the above comment isn't needed.

File: src/resources/views/columns/permissions.blade.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    $permissions = $entry->{$column['entity']};
+
+    // Groups permissions by prefix, sorted alphabetically
+    $permissionsByPrefix = collect($permissions)
+        ->sortBy(function($permission) {
+            return $permission->prefix() ?: PHP_INT_MAX; // Use PHP_INT_MAX as a little trick for sorting permissions without prefix at the end
```
Comment (lloy0076) [inline src/resources/views/columns/permissions.blade.php]: Ok, "alphabetical" and PHP_INT_MAX cause me to have mental dissonance.

What is that return comparing against (or is there some link explaining why it works)?

File: src/resources/views/columns/permissions.blade.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
+{{-- relationships with pivot table (n-n) --}}
+<td>
+    <?php
```
Comment (lloy0076) [inline src/resources/views/columns/permissions.blade.php]: `@php`

File: src/resources/views/columns/permissions.blade.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    // Groups permissions by prefix, sorted alphabetically
+    $permissionsByPrefix = collect($permissions)
+        ->sortBy(function($permission) {
+            return $permission->prefix() ?: PHP_INT_MAX; // Use PHP_INT_MAX as a little trick for sorting permissions without prefix at the end
+        })
+        ->groupBy(function($permission) {
+            return $permission->prefix();
+        });
+    ?>
```
Comment (lloy0076) [inline src/resources/views/columns/permissions.blade.php]: `@endphp`

File: src/resources/views/columns/permissions.blade.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                    ,
+                @else
+                    <br />
+                @endif
+            @endif
+
+            {{ $prefix }}
+
+            <em>
+                ({{ $permissions->map(function($permission) use ($attribute) {
```
Comment (lloy0076) [inline src/resources/views/columns/permissions.blade.php]: What does this do? (i.e. this probably could use a comment)

File: src/resources/views/fields/permissions.blade.php [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                </div>
+                <div class="col-sm-7">
+                    @foreach ($permissions as $permission)
+                        <div class="checkbox inline no-margin">
+                            <label>
+                                <input
+                                    type="checkbox"
+                                    name="{{ $field['name'] }}[]"
+                                    value="{{ $permission->getKey() }}"
+                                    @if( ( old( $field["name"] ) && in_array($permission->getKey(), old( $field["name"])) ) || (isset($field['value']) && in_array($permission->getKey(), $field['value']->pluck($permission->getKeyName(), $permission->getKeyName())->toArray())))
```
Comment (lloy0076) [inline src/resources/views/fields/permissions.blade.php]: I'd be inclined to reformat that so it's more readable.

File: src/resources/views/fields/permissions.blade.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                                        checked = "checked"
+                                    @endif > {!! $permission->item() !!} &nbsp;
```
Comment (lloy0076) [inline src/resources/views/fields/permissions.blade.php]: Why is this marked with `{!! !!}` here but above the roles/etc are escaped?

File: src/resources/views/fields/permissions.blade.php [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                            <i class="fa fa-check-square-o"></i>&nbsp; All
+                        </button>
+                    </div>
+                </div>
+        </div>
+    @endforeach
+
+    {{-- HINT --}}
+    @if (isset($field['hint']))
+        <p class="help-block">{!! $field['hint'] !!}</p>
```
Comment (lloy0076) [inline src/resources/views/fields/permissions.blade.php]: esc vs unescaped?

File: src/resources/views/fields/permissions_with_roles.blade.php [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+                    @if ($permissionWithPrefixExists)
+                        <div class="col-sm-3">
+                            <label class="no-margin">
+                                <strong>{{ $prefix }}</strong>
+                            </label>
+                        </div>
+                    @endif
+                    <div class="col-sm-{{ $permissionWithPrefixExists ? 7 : 12 }}">
+                        @foreach ($permissions as $permission)
+                            <?php
```
Comment (lloy0076) [inline src/resources/views/fields/permissions_with_roles.blade.php]: `@php`

File: src/resources/views/fields/permissions_with_roles.blade.php [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+                            $value = array_get($field, 'value');
+                            $hasPermissionViaUser = ($value[1]->pluck('id')->contains($permission->id)) || (old($fieldPermission['name']) && in_array($permission->id, old($fieldPermission['name'])));
+                            $hasPermissionViaRole = $entityRolesPermissions->contains($permission->id);
+                            ?>
```
Comment (lloy0076) [inline src/resources/views/fields/permissions_with_roles.blade.php]: `@endphp`

File: src/resources/views/fields/permissions_with_roles.blade.php [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+                            <div class="checkbox inline no-margin">
+                                <label>
+                                    <input
+                                        type="checkbox"
+                                        class="secondary_list"
+                                        data-id="{{ $permission->id }}"
+                                        @foreach ($fieldPermission as $attribute => $value)
+                                            @if (is_string($attribute) && $attribute != 'value' && !is_callable($value))
+                                                @if ($attribute=='name')
```
Comment (lloy0076) [inline src/resources/views/fields/permissions_with_roles.blade.php]: Spaces around the `==`

File: src/resources/views/fields/permissions_with_roles.blade.php [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+                                        {{ $permission->item() }}
+                                    @endif
+                                </label>
+                            </div>
+                        @endforeach
+                    </div>
+                    @if ($permissionWithPrefixExists)
+                        <div class="col-sm-2">
+                            <div class="pull-right">
+                                <button href="" class="btn btn-default btn-xs uncheck-row" title="Uncheck all" class="">
```
Comment (lloy0076) [inline src/resources/views/fields/permissions_with_roles.blade.php]: `Uncheck all` i18n

File: src/resources/views/fields/permissions_with_roles.blade.php [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+                                    <i class="fa fa-square-o"></i>&nbsp; None
+                                </button>
+                                &nbsp;
+                                <button href="" class="btn btn-default btn-xs check-row" title="Check all">
```
Comment (lloy0076) [inline src/resources/views/fields/permissions_with_roles.blade.php]: `Check all` i18n

**Comment by tabacitu** [comment]
Hi @shaoshiva - this looks like a lot of good work, congrats! I wish you had asked before going through all this trouble... Unfortunately, I don't think we can merge this, because:
1) It's beyond the scope of Backpack\CRUD to provide fields for permissions, so I think having some fields in CRUD that are used only by another Backpack package is a bad precedent; if it's not a general field, I don't think it should be included in CRUD; PermissionManager now uses the ```checklist_dependecy``` field, which is a general one, for that sole reason - reusability; 
2) Not everybody uses Permissions that way; some people use a simple permission system (if you can read "news" you can do anything to them), because some projects just don't need a permission system this fine-grained; I, for one, am a big believer in not over-building - keeping things as simple as possible for the user/admin makes sure he doesn't get intimidated to use the admin panel (like most of my WordPress clients have been, because of the sheer complexity of it);
3) I totally agree on the ability to make some column unsearchable - thank you; we've introduced an option for that in Backpack\CRUD 3.3 (you can define ```'searchLogic' => false``` on any column; this has the added benefit of being able to give it a closure, to specify a custom search logic);

But **I do think a lot of people would be happy to be able to use this easily**. I think it's a pretty popular use case and we _should_ make it easy for people to use it this way. And I think the best way to use it would be in a separate package, that publishes the few extra fields it needs. So I think a great way forward would be to create a fork of PermissionManager, say... AdvancedPermissionManager, that works like this. What do you think @shaoshiva ? Would you like to create and maintain such a package? If not, anybody else?

I'd be more than happy for PermissionManager to have some competition. After all, we don't all have to use the standard, simple packages. And the more packages we have, the easier it is for all of us to create our admin panels. As long as they're maintained, that is, of course :-) So if you create a separate package for this, I'd be happy to put it on BackpackForLaravel.com and tell the world using the newsletter, so it gets more downloads and feedback.

Tell me what you think. If you need help getting the package setup you can usually find us online on [Gitter](https://gitter.im/BackpackForLaravel/Lobby). 

Thanks, cheers!

**Pull request closed without merging**